src/composer.ts

```diff
@@ -232,33 +335,108 @@ export class AtomicTransactionComposer {
         );
       }
 
-      appArgTypes.push(argSpec.type);
-      appArgValues.push(argValue);
+      if (abiTypeIsReference(argType)) {
+        refArgIndexToBasicArgIndex.set(
+          refArgTypes.length,
+          basicArgTypes.length
+        );
+        refArgTypes.push(argType);
+        refArgValues.push(argValue);
+        // treat the reference as a uint8 for encoding purposes
+        argType = new ABIUintType(8);
+      }
+
+      if (typeof argType === 'string') {
+        throw new Error(`Unknown ABI type: ${argType}`);
+      }
+
+      basicArgTypes.push(argType);
+      basicArgValues.push(argValue);
+    }
+
+    const resolvedRefIndexes: number[] = [];
+    const foreignAccounts: string[] = [];
+    const foreignApps: number[] = [];
+    const foreignAssets: number[] = [];
```

Comment:
One question, do we want to constrain the number of foreign asset/account/app number?
I remembered in @algochoi py-sdk that we defined some number of constraints.
```python
    # The maximum number of foreign objects that can be associated per app
    FOREIGN_ACCOUNT_LIMIT = 4
    FOREIGN_ARRAY_LIMIT = 8
```
Is it needed here in a similar sense?

Reply:
My point of view is that these should not be checked in the SDKs because:
* These are consensus params that are liable to change in the future, and if they do we'll need to quickly update the SDKs with the new values
* If too many values are included for one of these arrays, algod _should_ provide a sufficiently clear error message when you submit the transaction, so not much is gained by raising the error before submission
* The existing application call transaction types in each SDK do _not_ currently check these limits, so I don't think we should do checking at the AtomicTransactionComposer level either

Reply:
Got it. If my understanding is correct, we might leave this to be handled on the algod side (which is 1 or somewhat more levels lower than AtomicTransactionComposer).